packages/xo-server/src/xapi-stats.js

```diff
@@ -516,70 +286,96 @@ export default class XapiStats {
       // timestamp of the oldest data value
       // So, we use the timestamp of the oldest data value !
       const startTimestamp = json.data[json.meta.rows - 1].t
-
-      // Remove useless data and reorder
-      // Note: Older values are at end of json.data.row
-      const parseOffset =
-        (this._hosts[hostname][step].endTimestamp - startTimestamp + step) /
-        step
-
-      json.data.splice(json.data.length - parseOffset)
-      json.data.reverse()
+      const endTimestamp = get(this._statsByObject, [
+        hostUuid,
+        step,
+        'endTimestamp',
+      ])
+
+      const statsOffset = endTimestamp - startTimestamp + step
+      if (endTimestamp !== undefined && statsOffset > 0) {
+        const parseOffset = statsOffset / step
+        // Remove useless data
+        // Note: Older values are at end of json.data.row
+        json.data.splice(json.data.length - parseOffset)
+      }
 
       // It exists useful data
       if (json.data.length > 0) {
-        const [hostLegends, vmsLegends] = parseLegends(json)
-
-        // Compute and update host/vms stats
-        this._parseVmsStats(json, hostname, vmsLegends, step)
-        this._parseHostStats(json, hostname, hostLegends, step)
-
-        // Remove older stats
-        this._removeOlderStats(
-          hostLegends,
-          this._hosts[hostname][step].stats,
-          RRD_POINTS_PER_STEP[step]
-        )
-
-        for (const uuid in vmsLegends) {
-          this._removeOlderStats(
-            vmsLegends[uuid],
-            this._vms[hostname][step][uuid],
-            RRD_POINTS_PER_STEP[step]
+        // reorder data
+        json.data.reverse()
+        forEach(json.meta.legend, (legend, index) => {
+          const [, type, uuid, metricType] = /^AVERAGE:([^:]+):(.+):(.+)$/.exec(
+            legend
           )
-        }
+
+          const metrics = STATS[type]
+          if (metrics === undefined) {
+            return
+          }
+
+          const { metric, testResult } = findMetric(metrics, metricType)
+
+          if (metric === undefined) {
+            return
+          }
+
+          const path =
+            metric.getPath !== undefined
+              ? metric.getPath(testResult)
+              : [findKey(metrics, metric)]
+
+          const metricValues = getValuesFromDepth(this._statsByObject, [
+            uuid,
+            step,
+            'stats',
+            ...path,
+          ])
+
+          metricValues.push(
+            ...computeValues(json.data, index, metric.transformValue)
+          )
+
+          // remove older Values
+          metricValues.splice(
+            0,
+            metricValues.length - RRD_POINTS_PER_STEP[step]
+          )
+        })
+
+        forEach(this._statsByObject, obj => {
+          if (obj[step] !== undefined) {
+            this._parseMemory(obj[step].stats)
+          }
+        })
```

Comment:
I still feel uneasy with this virtual metric. Do you think it would be easy to fix its uses to rely on `total - free` instead?

Reply:
Yes, it's better. I will do the necessary changes in `xo-usage-report` and `xo-web`.